cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+
+	pattern := fmt.Sprintf("%s/dev/block/%d:*", dirs.SysfsDir, d.major)
+	allDevices, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+
+	// glob does not guarantee a sort, but we need our list of devices sorted
+	sort.Strings(allDevices)
+
+	// populate all the partitions from our candidate devices
+	for _, dev := range allDevices {
+		base := filepath.Base(dev)
+		maj, min, err := parseDeviceMajorMinor(base)
+		if err != nil {
+			continue
+		}
+
+		// ignore any devices that have minor numbers less than the disk itself
+		// the disk will have the same minor so ignore that too
+		if min <= d.minor {
+			continue
+		}
+
+		// now if there is a partition file, this is a partition for our disk
+		if osutil.FileExists(filepath.Join(dev, "partition")) {
+			// read the uevent file for this partition to get the devname
+			p := &partition{
+				major: maj,
+				minor: min,
+			}
+			f, err := os.Open(filepath.Join(dev, "uevent"))
+			if err != nil {
+				continue
+			}
+
+			// now get the full set of udev properties for this partition
+			eventProps, err := parseUdevProperties(f)
+			if err != nil {
+				continue
+			}
+
+			// get the name of the device path to call udevadm
+			// note that here DEVNAME is coming from sysfs's uevent, and for
+			// whatever reason it is just the basename a la vda3, but DEVNAME
+			// when  requested from udevadm proper is the full path a la
+			// /dev/vda3
+			if name, ok := eventProps["DEVNAME"]; ok {
+				p.path = filepath.Join("/dev", name)
+			} else {
+				continue
+			}
+
+			props, err := udevProperties(p.path)
+			if err != nil && props == nil {
+				// only error here if we didn't get a map, we validate the map
+				// in the next steps
+				continue
+			}
+
+			// get the label
+			if labelEncoded, ok := props["ID_FS_LABEL_ENC"]; ok {
+				p.label = labelEncoded
+			} else {
+				continue
+			}
+
+			// finally get the partition uuid
+			if partuuid, ok := props["ID_PART_ENTRY_UUID"]; ok {
+				p.partuuid = partuuid
+			} else {
+				continue
+			}
+
+			d.partitions = append(d.partitions, p)
+		} else {
+			// if there was not a partition file, we hit another disk and must
+			// stop searching (the disk we are looking at will be ignored with
+			// the minor number <= check above)
+			break
+		}
+	}
+
+	// if we didn't find any partitions from above then return an error
+	if len(d.partitions) == 0 {
+		return nil, fmt.Errorf("no partitions found for disk %s", majorMinor)
+	}
+
+	return d, nil
+}
+
+func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
+	// iterate over the partitions looking for the specified label
+	for _, part := range d.partitions {
+		if part.label == label {
+			return part.partuuid, nil
+		}
+	}
+
+	return "", fmt.Errorf("couldn't find label %q", label)
+}
+
+func (d *disk) MountPointIsFromDisk(mountpoint string, opts *Options) (bool, error) {
+	d2, err := diskFromMountPointImpl(mountpoint, opts)
```

Comment:
sorry, GH lost of my comments.  I see now that the mocking is using an indepent implemetation of this. Its a a bit unclear why we care for this to use the real one under that scenario though. It would be best to avoid the Wrapper stuff, it's fairly confusing to the causal reader.

Reply:
well what I was trying to avoid here is the fact that in MountPointIsFromDisk and in DiskFromMountPoint in the real code, we want both of these things to use the same implementation of using sysfs + udev, but the signature on `diskFromMountPointImpl` returns a `*disk`, but `DiskFromMountPoint` returns a `Disk`, so I can't just assign `diskFromMountPointImpl` to `DiskFromMountPoint`, and if I make `diskFromMountPointImpl` return a `Disk`, then I need to use a type-assertion on it to get a `Disk` and my history of trying to use type assertions is that they are usually nack'd. 

I'm happy to get rid of the impl + wrapper things if you are ok to ack using a type assertion/switch :smile: otherwise I don't know how to satisfy both requirements (no type switches and no confusing wrapper/impl functions)

Reply:
I see, now I understand, I think giving it a name and a comment makes it more confusing than just doing it, I recommend just doing this,   in it own var stanza at the top:

```
var  diskFromMountPoint = func(mountpoint string, opts *Options) (Disk, error) {
	return diskFromMountPointImpl(mountpoint, opts)
}
```

no Wrapper things explicitly named

Reply:
ok, yes that sounds fine